src/emissions/emission.py

```diff
@@ -456,55 +348,568 @@ def get_GSE_emissions(self, wnsf):
+        elif method == 'none':
+            PMnvolEI = np.zeros_like(fuel_flows_LTO)
+            PMnvolEIN = None
+        else:
+            raise ValueError(
+                f"""Re-define PMnvol estimation method:
+                pmnvolSwitch = {self.pmnvol_method}"""
+            )
+
+        self.LTO_emission_indices['PMnvol'] = PMnvolEI
+        if (
+            self._include_pmnvol_number
+            and PMnvolEIN is not None
+            and 'PMnvolN' in self.LTO_emission_indices.dtype.names
+        ):
+            self.LTO_emission_indices['PMnvolN'] = PMnvolEIN
+
+    def _wnsf_index(self, wnsf: str) -> int:
+        """Map user-friendly WNSF labels to the internal lookup order."""
+        mapping = {'wide': 0, 'narrow': 1, 'small': 2, 'freight': 3}
+        idx = mapping.get(wnsf.lower())
+        if idx is None:
+            raise ValueError(
+                "Invalid WNSF code; must be one of 'wide','narrow','small','freight'"
+            )
+        return idx
+
+    def _gse_nominal_profile(self, idx: int):
+        """Return nominal per-cycle GSE emissions for the requested aircraft class."""
+        return {
+            'CO2': [58e3, 18e3, 10e3, 58e3][idx],
+            'NOx': [0.9e3, 0.4e3, 0.3e3, 0.9e3][idx],
+            'HC': [0.07e3, 0.04e3, 0.03e3, 0.07e3][idx],
+            'CO': [0.3e3, 0.15e3, 0.1e3, 0.3e3][idx],
+            'PM10': [0.055e3, 0.025e3, 0.020e3, 0.055e3][idx],
+        }
+
+    def _scope11_profile(self, ac_performance):
+        """Cache SCOPE11 lookups so we do the work only once."""
+        if self._scope11_cache is None:
+            edb = ac_performance.EDB_data
+            mass = calculate_PMnvolEI_scope11(
+                np.array(edb['SN_matrix']),
+                np.array(edb['PR']),
+                np.array(edb['ENGINE_TYPE']),
+                np.array(edb['BP_Ratio']),
+            )
+            number = edb.get('PMnvolEIN_best_ICAOthrust')
+            self._scope11_cache = {
+                'mass': np.asarray(mass, dtype=float),
+                'number': None if number is None else np.asarray(number, dtype=float),
+            }
+        return self._scope11_cache
+
+    def _map_mode_values_to_categories(
+        self, mode_values: np.ndarray, thrust_categories: np.ndarray
+    ):
+        """Broadcast mode-level EIs to arbitrary thrust-category time series."""
+        values = np.asarray(mode_values, dtype=float).ravel()
+        if values.size == 0:
+            return np.zeros_like(thrust_categories, dtype=float)
+        mapped = np.full(thrust_categories.shape, values[-1], dtype=float)
+        mapped[thrust_categories == 2] = values[0]
+        if values.size > 1:
+            mapped[thrust_categories == 3] = values[1]
+        if values.size > 2:
+            mapped[thrust_categories == 1] = values[2]
+        return mapped
+
     def __emission_dtype(self, shape):
+        """Build the structured dtype used for every emission array."""
```

Comment:
I still don't like this approach very much! We've talked about this one before though, I think.